UML-Diagram/ReportSystem.puml

```diff
@@ -1,32 +1,28 @@
 @startuml ReportSystem
-package "Report System" {
-    
-    ' Define main abstractc
-    abstract class ABorrowReport implements IReporting{
-        - BorrowManager : BorrowList
-    }
 
-    abstract class APatronReport implements IReporting{
-        - PatronManager : PatronList
+package "Report System" {
+    ' Define main contract
+    interface IReport {
+        + GenerateReport() : void
     }
 
-    ' Define main interface
-    interface IReporting {
-        - LogService : ILogService
-        + generateReport()
+    ' Define main abstract base
+    abstract class BaseReport implements IReport {
+        - BorrowManager : _borrowManager
     }
 
     ' Define main classes
-    class BookCurrentBorrwedReport extends ABorrowReport {
-        + BookBorrowedReportFormat()
+    class BorrowPatronHistory extends BaseReport {
+        + PatronBorrowHistoryReport() : void
```

Comment:
I think that its missing a client class wich you can call and set the kind of BorrowReport and an execute method to  display the report, but whay you think

Reply:
We can create a new class as you suggest but it won't be the type of client. It should be a concrete for borrows client classes.
But if we review our actual code on **develop** branch, `BookReport` and `PatronReport` classes, first of all implements `Reporting` interface, which only has the next function.
```
interface Reporting
{
    public void generateReport();
}
```
So, `BookReport` and `PatronReport` classes have the next code.
```
public abstract class BookReport: Reporting
{
    protected BorrowManager BorrowList;
    public BookReport( BorrowManager BorrowList)
    {
        this.BorrowList = BorrowList;
    }

    public void generateReport()
    {
        Console.WriteLine("This is a book borrow Report");
    }
}
```
```
public abstract class PatronReport : Reporting
{
    protected BorrowManager Borrows;
    public PatronReport(BorrowManager Borrows)
    {
        this.Borrows = Borrows;
    }

    public void generateReport()
    {
        Console.WriteLine("This is a Patron borow Report");
    }
}
```
As result, our client classes will have the next code. Where concrete class share its attribute and let client class access the `generateReport` method and its content, which comes from the contract.
```
public class BorrowPatronHistory : PatronReport
{
    public BorrowPatronHistory(BorrowManager BorrowList) : base(BorrowList){}

    public void PatronBorrowHistoryReport(string PatronName)
    {
        generateReport();
        foreach(Borrow borrow in Borrows.GetBorrows()){
            if(borrow.GetPatron().getName() == PatronName){
                borrow.BorrowDetails();
            }
        }        
    }
}
```
To sum up, if we're still creating concretes for specific client classes, we'll have one for everyone. Having as result redundancy classes and violating **DRY** principle instead of making of use a general or base concrete class for client classes.

Reply:
We can create a new class as you suggest but it won't be the type of client. It should be a concrete for borrows client classes.
But if we review our actual code on **develop** branch, `BookReport` and `PatronReport` classes, first of all implements `Reporting` interface, which only has the next function.
```
interface Reporting
{
    public void generateReport();
}
```
So, `BookReport` and `PatronReport` classes have the next code.
```
public abstract class BookReport: Reporting
{
    protected BorrowManager BorrowList;
    public BookReport( BorrowManager BorrowList)
    {
        this.BorrowList = BorrowList;
    }

    public void generateReport()
    {
        Console.WriteLine("This is a book borrow Report");
    }
}
```
```
public abstract class PatronReport : Reporting
{
    protected BorrowManager Borrows;
    public PatronReport(BorrowManager Borrows)
    {
        this.Borrows = Borrows;
    }

    public void generateReport()
    {
        Console.WriteLine("This is a Patron borow Report");
    }
}
```
As result, our client classes will have the next code. Where concrete class share its attribute and let client class access the `generateReport` method and its content, which comes from the contract.
```
public class BorrowPatronHistory : PatronReport
{
    public BorrowPatronHistory(BorrowManager BorrowList) : base(BorrowList){}

    public void PatronBorrowHistoryReport(string PatronName)
    {
        generateReport();
        foreach(Borrow borrow in Borrows.GetBorrows()){
            if(borrow.GetPatron().getName() == PatronName){
                borrow.BorrowDetails();
            }
        }        
    }
}
```
To sum up, if we're still creating concretes for specific client classes, we'll have one for everyone. Having as result redundancy classes and violating **DRY** principle. That's why I suggest to use a general or base concrete class for client classes.

Reply:
That are the concrete classes, that are not the clients, BorrowPatronHistory its a conrete of PatronReport, and  i agree for merging both templates in one because has the same structure the PatronReport and BookReport, but again it will be usefull to have a client class that uses the interface has a paremeter in order to set new concretes of the new ReportTemplate and doesnt has to initialize every class. Use the reference to the family/group of methods the interface in order to only instance one class that can set the different types of concretes that are using the interface. Something like:

public class ReportClient
+ SetReport (IReport report) : void -> in this method set the type of Report
+ ExecuteReport () : void -> this method execute the method of the seted Report

Reply:
What do you mean by "initialize every class"? Are you talking about `BorrowManager` instance of `BaseReport` class?

Reply:
Are you we really need another class to do that? With the use of IReport in some class, we already can do that. So I don't think we need one class to use the concretes if we can do it with the contract.

Reply:
It can be, it will work, but that doesnt reflect in the UML, IReport doesnt have a relation from his use in other class